gh-126399 minor changes in RawTurtle.clone() method

Chaebin-Kim24 opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->

This is a minor change of RawTurtle.clone() method to correct self.items of the cloned Turtle. 

In clone() method, regarding items and currentLineItem,
- First, self.currentLineItem is updated to a new one, and the item is appended to self.items by `self._newLine(self._drawing)`
- Then, deepcopy of self is performed by `q = self.deepcopy()` 
- Finally, q.currentLineItem is updated to a new one by `q.currentLineItem = screen._createline()`

Potential problems of this code is that
- updated self.currentLineItem is in q.items, so q.clear() deletes the new self.currentLineItem
- q.currentLineItem is not in q.items, so q.clear() never deletes the q.currentLineItem 

To address these issues, following changes were made.
- update self.currentLineItem after performing deepcopy
- append q.currentLineItem to q.items

<!-- gh-issue-number: gh-126399 -->
* Issue: gh-126399
<!-- /gh-linked-prs -->
* Related PR: gh-127189

Chaebin-Kim24 committed:
Update turtle.py

Update RawTurtle.clone() method.

1) moved self._newLine(self._drawing) after deepcopy()
2) added q.items.append(q.currentLineItem) after q.currentLineItem = screen._createline()

Reason for 1):
When clear method is called on the cloned Turtle object, the method also deletes the first line of the original Turtle object which is drawn after cloning.

Reason for 2):
When clear method is called on the cloned Turtle object, the method does not delete the first line of the cloned Turtle object which is drawn after cloning

Chaebin-Kim24 committed:
Merge branch 'python:main' into main

Chaebin-Kim24 committed:
Merge branch 'main' into main

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

Zheaoli commented on Misc/NEWS.d/next/Library/2024-11-04-15-08-02.gh-issue-126399.-4rB22.rst:
```diff
@@ -0,0 +1 @@
+Fix the turtle module
```
```suggestion
Fix :mod:`turtle` clone() items management when clearing cloned turtle items
```

Chaebin-Kim24 committed:
Update Misc/NEWS.d/next/Library/2024-11-04-15-08-02.gh-issue-126399.-4rB22.rst

Co-authored-by: Nadeshiko Manju <lizheao940510@gmail.com>

Chaebin-Kim24 committed:
Merge branch 'main' into main

Chaebin-Kim24 committed:
Merge branch 'main' into main

terryjreedy commented on Lib/turtle.py:
```diff
@@ -2812,13 +2812,13 @@ def clone(self):
         self.turtle = None  # too make self deepcopy-able
 
         q = deepcopy(self)
+        self._newLine(self._drawing)
```
The doctest fails with 20 failures: https://github.com/python/cpython/actions/runs/11687342848/job/32545252133?pr=126401.  One may have to click the label "Search log" label to see the log.  This is the first failure.

File "library/turtle.rst", line 1688, in default
Failed example:
    joe = mick.clone()
Exception raised:
    Traceback (most recent call last):
      File "/home/runner/work/cpython/cpython/Lib/doctest.py", line 1395, in __run
        exec(compile(example.source, filename, "single",
        ~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                     compileflags, True), test.globs)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "<doctest default[1]>", line 1, in <module>
        joe = mick.clone()
      File "/home/runner/work/cpython/cpython/Lib/turtle.py", line 2821, in clone
        self._newLine(self._drawing)
        ~~~~~~~~~~~~~^^^^^^^^^^^^^^^
      File "/home/runner/work/cpython/cpython/Lib/turtle.py", line 3367, in _newLine
        self.screen._drawline(self.currentLineItem, top=True)
        ^^^^^^^^^^^^^^^^^^^^^
    AttributeError: 'NoneType' object has no attribute '_drawline'

This failure seems to be caused by moving _drawline() to a location where self.screen has been set to None.  So I suspect that this line should be after self.screen is restored.

The remaining failures might be side-effects of this failure, but I don't know how doctest works.
@AA-Turner You recently patched turtle doctests.  Do you have any better insight?

terryjreedy replied:
Line was moved back to whee it was and doctest passes.

terryjreedy reviewed (changes requested):
1. Fix the doctest failure (see comment).  DONE
2. Add a unittest that fails without the patch and passes with it.

EDIT: I thought of expanding the doc clone example to make it a true test, but it might require too much for readers.

I am not very familiar either with doctest or the turtle code.  So not likely to make a final decision and merge.

Chaebin-Kim24 committed:
Merge branch 'main' into main

Chaebin-Kim24 committed:
Merge branch 'python:main' into main

Chaebin-Kim24 committed:
Update turtle.py

moved self._newLine(self._drawing) to original position

changed q.items.append(q.currentLineItem) to q.items[-1] = q.currentLineItem

terryjreedy commented on Lib/turtle.py:
```diff
@@ -2819,7 +2819,7 @@ def clone(self):
 
         q = deepcopy(self)
-
+        
```
```suggestion
```

terryjreedy committed:
Update Lib/turtle.py

terryjreedy commented on Lib/turtle.py:
```diff
@@ -2819,7 +2819,6 @@ def clone(self):
         self.turtle = None  # too make self deepcopy-able
 
         q = deepcopy(self)
```
```suggestion
        q = deepcopy(self)

```

terryjreedy committed:
Update Lib/turtle.py

Chaebin-Kim24 committed:
Merge branch 'main' into main

Chaebin-Kim24 committed:
Update turtle.rst

added a unit test that fails without the patch and passes with it

Chaebin-Kim24 committed:
Update turtle.rst

revert change

Chaebin-Kim24 committed:
Update test_turtle.py

Added a unittest that fails without the patch and passes with it.

Chaebin-Kim24 committed:
Merge branch 'python:main' into main

Chaebin-Kim24 committed:
Update test_turtle.py

TestRawTurtle.test_clone():

copied screen from TestTurtleScreen.test_save_raises_if_wrong_extension

Chaebin-Kim24 committed:
Update test_turtle.py

TestRawTurtle.test_clone():

screen = Screen() as in Turtle class

Chaebin-Kim24 committed:
Merge branch 'python:main' into main

Chaebin-Kim24 committed:
Update test_turtle.py

Screen belongs to turtle namespace

Chaebin-Kim24 committed:
Update test_turtle.py

check variable name

Chaebin-Kim24 committed:
Merge branch 'python:main' into main

Chaebin-Kim24 committed:
Update test_turtle.py

handling _tkinter.TclError

Chaebin-Kim24 committed:
Update test_turtle.py

correct qualified name of TclError

Chaebin-Kim24 committed:
Merge branch 'main' into main

Chaebin-Kim24 committed:
Update test_turtle.py

raise SkipTest to skip test instead of skipTest()

Chaebin-Kim24 committed:
Update test_turtle.py

correct qualified name of unittest.case.SkipTest exception class

Chaebin-Kim24 committed:
Merge branch 'python:main' into main

Chaebin-Kim24 committed:
Merge branch 'main' into main

Chaebin-Kim24 committed:
Merge branch 'main' into main

Chaebin-Kim24 committed:
Merge branch 'main' into main

Chaebin-Kim24 committed:
Merge branch 'python:main' into main

Chaebin-Kim24 committed:
Merge branch 'main' into main

Chaebin-Kim24 committed:
Update test_turtle.py

lint: remove trailing whitespaces

Chaebin-Kim24 committed:
Merge branch 'main' into main

Chaebin-Kim24 committed:
Merge branch 'main' into main

Chaebin-Kim24 committed:
Merge branch 'python:main' into main

Chaebin-Kim24 committed:
Merge branch 'python:main' into main

hugovk commented on Lib/test/test_turtle.py:
```diff
@@ -465,6 +465,19 @@ def test_teleport(self):
+        rawturtle = turtle.RawTurtle(self.screen)
+        another_turtle = rawturtle.clone()
+        self.assertEqual(another_turtle.currentLineItem, another_turtle.items[-1])
+        self.assertFalse(rawturtle.currentLineItem in another_turtle.items)
```
Let's use underscores, and also [`assertNotIn`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertNotIn):
```suggestion
        raw_turtle = turtle.RawTurtle(self.screen)
        another_turtle = raw_turtle.clone()
        self.assertEqual(another_turtle.currentLineItem, another_turtle.items[-1])
        self.assertNotIn(raw_turtle.currentLineItem, another_turtle.items)
```

hugovk commented on Misc/NEWS.d/next/Library/2024-11-04-15-08-02.gh-issue-126399.-4rB22.rst:
```diff
@@ -0,0 +1 @@
+Fix :mod:`turtle` clone() items management when clearing cloned turtle items
```
We can link directly to the method:
```suggestion
Fix :meth:`turtle.clone` items management when clearing cloned turtle items.
```

Chaebin-Kim24 committed:
Update Lib/test/test_turtle.py

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

Chaebin-Kim24 committed:
Update Misc/NEWS.d/next/Library/2024-11-04-15-08-02.gh-issue-126399.-4rB22.rst

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>
